compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2202,8 +2217,138 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
     }
 
-    /// Literals and const generic parameters are eagerly converted to a constant, everything else
-    /// becomes `Unevaluated`.
+    #[instrument(level = "debug", skip(self), ret)]
+    pub fn lower_const_assoc_path(
+        &self,
+        hir_ref_id: HirId,
+        span: Span,
+        qself_ty: Ty<'tcx>,
+        qself: &'tcx hir::Ty<'tcx>,
+        assoc_segment: &'tcx hir::PathSegment<'tcx>,
+    ) -> Const<'tcx> {
+        debug!(%qself_ty, ?assoc_segment.ident);
+        let tcx = self.tcx();
+
+        let assoc_ident = assoc_segment.ident;
+
+        // Check if we have an enum variant or an inherent associated const.
+        // FIXME(mgca): handle assoc fns once we support those
+        if let Some(adt_def) = self.probe_adt(span, qself_ty) {
+            if adt_def.is_enum() {
+                let variant_def = adt_def
+                    .variants()
+                    .iter()
+                    .find(|vd| tcx.hygienic_eq(assoc_ident, vd.ident(tcx), adt_def.did()));
+                if let Some(variant_def) = variant_def {
+                    tcx.check_stability(variant_def.def_id, Some(hir_ref_id), span, None);
+                    let _ = self.prohibit_generic_args(
+                        slice::from_ref(assoc_segment).iter(),
+                        GenericsArgsErrExtend::EnumVariant { qself, assoc_segment, adt_def },
+                    );
+                    let uv = ty::UnevaluatedConst::new(variant_def.def_id, ty::List::empty());
+                    return Const::new_unevaluated(tcx, uv);
+                }
+            }
+
+            // FIXME(mgca): Support self types other than ADTs.
+            let candidates = tcx
+                .inherent_impls(adt_def.did())
+                .iter()
+                .filter_map(|&impl_| {
+                    self.probe_assoc_item(
+                        assoc_ident,
+                        ty::AssocKind::Const,
+                        hir_ref_id,
+                        span,
+                        impl_,
+                    )
+                    .map(|assoc| (impl_, assoc))
+                })
+                .collect::<Vec<_>>();
+            match &candidates[..] {
+                [] => {}
+                &[(impl_, assoc)] => {
+                    // FIXME(mgca): adapted from temporary inherent assoc ty code that may be incorrect
```

Comment:
same here we should just make `probe_inherent_assoc_ty` be able to be reused for consts somehow

Reply:
Done.